src/m365/spfx/commands/project/project-azuredevops-pipeline-add.ts

```diff
@@ -0,0 +1,213 @@
+        }
+
+        return true;
+      }
+    );
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  public async commandAction(logger: Logger, args: CommandArgs): Promise<void> {
+    this.projectRootPath = this.getProjectRoot(process.cwd());
+    if (this.projectRootPath === null) {
+      throw new CommandError(`Couldn't find project root folder`, SpfxProjectAzureDevOpsPipelineAddCommand.ERROR_NO_PROJECT_ROOT_FOLDER);
+    }
+
+    const solutionPackageJsonFile: string = path.join(this.projectRootPath, 'package.json');
+    const packageJson: string = fs.readFileSync(solutionPackageJsonFile, 'utf-8');
+    const solutionName = JSON.parse(packageJson).name;
+
+    if (this.debug) {
+      logger.logToStderr(`Adding Azure DevOps pipeline in the current SPFx project`);
+    }
+
+    try {
+      this.updatePipeline(solutionName, pipeline, args.options);
+      this.savePipeline(pipeline);
+    }
+    catch (error: any) {
+      throw new CommandError(error);
+    }
+  }
+
+  private savePipeline(pipeline: azureDevOpsPipeline): void {
+    const azureDevOpsPath: string = path.join(this.projectRootPath as string, '.azuredevops');
+    fsUtil.ensureDirectory(azureDevOpsPath);
+
+    const pipelinesPath: string = path.join(azureDevOpsPath, 'pipelines');
+    fsUtil.ensureDirectory(pipelinesPath);
+
+    const pipelineFile: string = path.join(pipelinesPath, 'deploy-spfx-solution.yml');
+    fs.writeFileSync(path.resolve(pipelineFile), yaml.stringify(pipeline), 'utf-8');
+  }
+
+  private updatePipeline(solutionName: string, pipeline: azureDevOpsPipeline, options: GlobalOptions): void {
+    if (options.name) {
+      pipeline.name = options.name;
+    }
+    else {
+      delete pipeline.name;
+    }
+
+    if (options.branchName) {
+      pipeline.trigger.branches.include[0] = options.branchName;
+    }
+
+    const version = this.getProjectVersion();
+
+    if (!version) {
+      throw `Unable to determine the version of the current SharePoint Framework project`;
+    }
+
+    const match = version.match(/^(\d+)\.(\d+)\.(\d+)$/);
+    const minorVersionString = match ? match[2] : null;
+    const minorVersion = minorVersionString ? Number(minorVersionString) : null;
+
+    if (minorVersion === null || isNaN(minorVersion)) {
+      throw `Unable to determine the minor version of the current SharePoint Framework project`;
+    }
+
+    if (minorVersion < 18) {
+      this.getNodeAction(pipeline).inputs!.versionSpec = '16.x';
+    }
+
+    if (options.loginMethod === 'user') {
+      const script = this.getScriptAction(pipeline);
+      script.script = script.script!.replace(`m365 login --authType certificate --certificateBase64Encoded '$(CertificateBase64Encoded)' --password '$(CertificatePassword)' --appId '$(EntraIDAppId)' --tenant '$(TenantId)'`, `m365 login --authType password --userName '$(UserName)' --password '$(Password)'`);
```

Comment:
If we suppress script as non-nullable, why do we define it as nullable in our type in the first place?

Reply:
in general we may have a ADO task that does not have a `script` option defined and in our case this is the option we have. As our type needs to be ready for a case in which we have a task with no script then I made it nullable. Here I suppress it as I am 'sure' we will have it in this place as in the previous line we get the action with the script option.
we may do a `if not null` condition here instead of suppress. what do you think?

Reply:
> As our type needs to be ready for a case in which we have a task with no script then I made it nullable

Is it a theoretical possibility? We're the only consumer of this code so we could base the types of what we actually need and change them in the future. Since it's all internal code, it wouldn't be a breaking change.

Reply:
it's the same case like in the comment above

Reply:
See reply. Let's add a null check to our code.